tasks/rule_generator/src/main.rs

```diff
@@ -0,0 +1,178 @@
+
+use convert_case::{Case, Casing};
+use oxc_allocator::Allocator;
+use oxc_ast::{
+    ast::{Argument, Expression, ObjectProperty, PropertyKey, PropertyValue},
+    AstKind, SourceType,
+};
+use oxc_parser::Parser;
+use oxc_semantic::SemanticBuilder;
+use serde::Serialize;
+
+const RULE_TEMPLATE: &str = include_str!("../template.txt");
+const ESLINT_TEST_PATH: &str =
+    "https://raw.githubusercontent.com/eslint/eslint/main/tests/lib/rules";
+
+#[derive(Debug)]
+enum TestCase<'a> {
+    Invalid(Cow<'a, str>),
+    Valid(Cow<'a, str>),
+}
+
+impl<'a> TestCase<'a> {
+    fn to_code(test_case: &TestCase) -> String {
+        match test_case {
+            TestCase::Valid(code) | TestCase::Invalid(code) => format!(r#"("{code}", None)"#),
+        }
+    }
+}
+
+#[derive(Serialize)]
+struct Context<'a> {
+    rule: &'a str,
+    pass_cases: &'a str,
+    fail_cases: &'a str,
+}
+
+fn parse_test_code<'a>(expr: &'a Expression) -> Option<Cow<'a, str>> {
+    match expr {
+        Expression::StringLiteral(lit) => Some(Cow::Borrowed(lit.value.as_str())),
+        Expression::TemplateLiteral(lit) => Some(Cow::Borrowed(lit.quasi().unwrap().as_str())),
+        Expression::ObjectExpression(obj_expr) => {
+            for obj_prop in &obj_expr.properties {
+                match obj_prop {
+                    ObjectProperty::Property(prop) => match &prop.key {
+                        PropertyKey::Identifier(ident) if ident.name == "code" => match &prop.value
+                        {
+                            PropertyValue::Expression(expr) => return parse_test_code(expr),
+                            PropertyValue::Pattern(_) => continue,
+                        },
+                        _ => continue,
+                    },
+                    ObjectProperty::SpreadProperty(_) => continue,
+                }
+            }
+            None
+        }
+        Expression::CallExpression(call_expr) => match &call_expr.callee {
+            Expression::MemberExpression(member_expr) => match &member_expr.object() {
+                // ['class A {', '}'].join('\n')
+                Expression::ArrayExpression(array_expr) => {
+                    let mut code = String::new();
+                    for arg in &array_expr.elements {
+                        let Some(Argument::Expression(Expression::StringLiteral(lit))) = arg else { continue };
+                        code.push_str(lit.value.as_str());
+                        code.push('\n');
+                    }
+                    Some(Cow::Owned(code))
+                }
+                _ => None,
+            },
+            _ => None,
+        },
+        _ => None,
+    }
+}
+
+#[tokio::main]
+async fn main() -> Result<(), reqwest::Error> {
+    let mut args = std::env::args();
+    let _ = args.next();
+    let rule_name = args.next().expect("expected rule name");
+    let rule_test_path = format!("{ESLINT_TEST_PATH}/{rule_name}.js");
```

Comment:
It's going to be difficult for rules that have tests from typescript-eslint, but I guess we'll have to copy manually at that point, because they have a lot of complicated stuff.

Reply:
For sure – we can expand on it as we go. Perhaps introducing a trait for different test suites?

```rs
struct TestSuite<'a> {
  pass: Vec<TestCase<'a>>,
  fail: Vec<TestCase<'a>>,
}

trait ParseTestSuite {
  fn parse(semantic: &Semantic) -> TestSuite;
}
```

and then we can write implementations for ESLint, `typescript-eslint`, etc?